### PR TITLE
Modify the relationship between forum, topic and post models

### PR DESCRIPTION
**Pull request by xzhflying**
* Modify the relationship between forum, topic and post models.
* Add some methods to calculate the statistic of these models.

### CODE REVIEW

File: app/controllers/course/forum/forums_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class Course::Forum::ForumsController < Course::ComponentController
   authorize_resource :forum, class: Course::Forum.name, through: :course
 
   def index
-    @forums = @forums.with_forum_statistics
```
Comment (lowjoel) [inline app/controllers/course/forum/forums_controller.rb]: You can preserve with_forum_statistics as a scope.

```ruby
scope :with_forum_statistics, ->() { calculated(:topic_count, :topic_view_count, :topic_post_count) }
```

File: app/models/course/discussion/topic.rb [hunk 2]
```diff
@@ -3,4 +3,6 @@ class Course::Discussion::Topic < ActiveRecord::Base
 
   has_many :posts, dependent: :destroy, inverse_of: :topic
   has_many :subscriptions, dependent: :destroy, inverse_of: :topic
+
+  accepts_nested_attributes_for :posts
```
Comment (lowjoel) [inline app/models/course/discussion/topic.rb]: Remember that if you are updating a topic you should not allow people to modify posts.

File: app/models/course/forum/topic.rb [hunk 3]
```diff
@@ -4,12 +4,44 @@ class Course::Forum::Topic < ActiveRecord::Base
 
   acts_as :topic, class_name: Course::Discussion::Topic.name, inverse_of: :actable
 
+  after_initialize :generate_initial_post
```
Comment (lowjoel) [inline app/models/course/forum/topic.rb]: `unless: persisted?` then remove the posts.empty check.

File: spec/models/course/forum/topic_spec.rb [hunk 4]
```diff
@@ -25,5 +26,57 @@
         end
       end
     end
+
+    describe '.generate_initial_post' do
```
Comment (lowjoel) [inline spec/models/course/forum/topic_spec.rb]: it's actually `#generate_initial_post`

File: spec/models/course/forum/topic_spec.rb [hunk 5]
```diff
@@ -25,5 +26,57 @@
...
+      let!(:topic) { create(:forum_topic, forum: forum, title: 'lol') }
+
+      context 'when create a new topic' do
```
Comment (lowjoel) [inline spec/models/course/forum/topic_spec.rb]: when creating

File: spec/models/course/forum/topic_spec.rb [hunk 6]
```diff
@@ -25,5 +26,57 @@
...
+
+        it 'does not change the number of posts and the title of the initial post' do
+          expect(topic.posts.count).to eq(1)
+          expect(topic.title).to eq('new title')
+          expect(topic.posts.first.title).to eq('lol')
+        end
+      end
+    end
+
+    describe 'Calculated :post_count' do
```
Comment (lowjoel) [inline spec/models/course/forum/topic_spec.rb]: just call it `.post_count`

**Comment by xzhflying** [comment]
@lowjoel

**Pull request merged**